frontend/test/resolution/testSplitInit.cpp

```diff
@@ -0,0 +1,860 @@
+  testSplitInit("test3",
+    R""""(
+      module M {
+        // this would be in the standard library...
+        operator =(ref lhs: int, rhs: int) {
+          __primitive("=", lhs, rhs);
+        }
+
+        proc test() {
+          var yes1:int;
+          yes1 = 1;
+        }
+      }
+    )"""",
+    {"yes1"});
+}
+
+
+static void test4() {
+  testSplitInit("test4",
+    R""""(
+      module M {
+        // this would be in the standard library...
+        operator =(ref lhs: int, rhs: int) {
+          __primitive("=", lhs, rhs);
+        }
+
+        proc test() {
+          var x:int = 0;
+          var yes2;
+          {
+            x = 24;
+            yes2 = 2;
+          }
+        }
+      }
+    )"""",
+    {"yes2"});
+}
+
+static void test5() {
+  testSplitInit("test5",
+    R""""(
+      module M {
+        // this would be in the standard library...
+        operator =(ref lhs: int, rhs: int) {
+          __primitive("=", lhs, rhs);
+        }
+
+        proc test(cond: bool) {
+          var x:int = 0;
+          var yes3;
+          var no: int;
+          {
+            if cond {
+              yes3 = 3;
+            } else {
+              x = 99;
+              x = 123;
+              {
+                yes3 = 3;
+              }
+            }
+          }
+          if cond {
+            no = 22;
+          }
+        }
+      }
+    )"""",
+    {"yes3"});
+}
+
+static void test6() {
+  testSplitInit("test6",
+    R""""(
+      module M {
+        // this would be in the standard library...
+        operator =(ref lhs: int, rhs: int) {
+          __primitive("=", lhs, rhs);
+        }
+
+        proc test(cond: bool, otherCond: bool) {
+          var yes5;
+          if cond {
+            yes5 = 5;
+          } else if otherCond {
+            yes5 = 55;
+          } else {
+            yes5 = 555;
+          }
+        }
+      }
+    )"""",
+    {"yes5"});
+}
+
+static void test7() {
+  testSplitInit("test7",
+    R""""(
+      module M {
+        // this would be in the standard library...
+        operator =(ref lhs: int, rhs: int) {
+          __primitive("=", lhs, rhs);
+        }
+
+        proc test() {
+          var no1 = 4;
+          no1 = 5;
+        }
+      }
+    )"""",
+    {});
+}
+
+static void test8() {
+  testSplitInit("test8",
+    R""""(
+      module M {
+        // this would be in the standard library...
+        operator =(ref lhs: int, rhs: int) {
+          __primitive("=", lhs, rhs);
+        }
+
+        proc test() {
+          {
+            var no2:int;
+            var tmp = no2;
+            no2 = 57;
+          }
+        }
+      }
+    )"""",
+    {});
+}
+
+static void test9() {
+  testSplitInit("test9",
+    R""""(
+      module M {
+        // this would be in the standard library...
+        operator =(ref lhs: int, rhs: int) {
+          __primitive("=", lhs, rhs);
+        }
+
+        proc test() {
+          {
+            var no3:int;
+            var tmp:no3.type;
```

Comment:
Hmm, I see nothing morally wrong with split-init'ing `no3` here (its type is used, not its value) but I don't think it's easy to write a sophisticated enough analysis.

Reply:
(That would be a language change, though, so we won't consider it here)